File: frontend/lib/types/shot.ts
export interface Shot {
  id: string;
  scene_id: string;
  number: number;
  shot_type: ShotType | null;
  camera_movement: CameraMovement | null;
  lighting: Lighting | null;
  colour_mood: ColourMood | null;
  action: string | null;
  dialogue: string | null;
  emotional_beat: string | null;
  estimated_duration_seconds: number;
  quality_tier: "wan" | "happyhorse" | "happyhorse_fast" | "deferred" | null;
  characters_in_frame: string[] | null;
  notes: string | null;
  director_note: string | null;
  blocking_json?: ShotBlocking | null;
}

/** Absolute per-shot geometry from the Director, 180-rule enforced. */
export interface ShotBlocking {
  subjects?: BlockingSubject[] | null;
  reverse_angle?: boolean;
}

export interface BlockingSubject {
  character?: string;
  frame_position?: "FG" | "MG" | "BG" | string;
  screen_side?: "left" | "center" | "right" | string;
  facing?: string;
  eyeline?: string;
  action?: string;
  posture?: "standing" | "sitting" | "lying" | "collapsed" | "kneeling" | string;
}

export type ShotType = "ECU" | "CU" | "MCU" | "MS" | "FS" | "LS" | "EWS" | "POV" | "OTS" | "INSERT";
export type CameraMovement = "STATIC" | "PAN_LEFT" | "PAN_RIGHT" | "TILT_UP" | "TILT_DOWN" | "DOLLY_IN" | "DOLLY_OUT" | "HANDHELD" | "DRONE";
export type Lighting = "NATURAL" | "GOLDEN_HOUR" | "BLUE_HOUR" | "NIGHT" | "OVERCAST" | "DRAMATIC_SIDE" | "NEON" | "PRACTICAL";
export type ColourMood = "WARM" | "COOL" | "DESATURATED" | "HIGH_CONTRAST" | "PASTEL" | "VIVID" | "MONOCHROME";

File: backend/app/schemas/shot.py
from pydantic import BaseModel
from uuid import UUID
from typing import Optional


class ShotCreate(BaseModel):
    scene_id: UUID
    number: int
    shot_type: Optional[str] = None
    camera_movement: Optional[str] = None
    lighting: Optional[str] = None
    colour_mood: Optional[str] = None
    action: Optional[str] = None
    dialogue: Optional[str] = None
    emotional_beat: Optional[str] = None
    estimated_duration_seconds: int = 5
    quality_tier: Optional[str] = None
    characters_in_frame: Optional[list[str]] = None
    notes: Optional[str] = None
    director_note: Optional[str] = None


class ShotResponse(BaseModel):
    id: UUID
    scene_id: UUID
    number: int
    shot_type: Optional[str] = None
    camera_movement: Optional[str] = None
    lighting: Optional[str] = None
    colour_mood: Optional[str] = None
    action: Optional[str] = None
    dialogue: Optional[str] = None
    emotional_beat: Optional[str] = None
    estimated_duration_seconds: int
    quality_tier: Optional[str] = None
    characters_in_frame: Optional[list[str]] = None
    notes: Optional[str] = None
    director_note: Optional[str] = None
    # absolute per-shot geometry: {"subjects": [{character, frame_position,
    # screen_side, facing, eyeline, action}], "reverse_angle": bool}
    blocking_json: Optional[dict] = None
    prompt_json: Optional[dict] = None

    model_config = {"from_attributes": True}
